pkg/clusterapi/controlplane_test.go

```diff
@@ -0,0 +1,322 @@
+	mt.Name = originalName
+	client := test.NewFakeKubeClient()
+
+	g.Expect(clusterapi.EnsureNewNameIfChanged(ctx, client, dummyRetriever, noChangesCompare, mt)).To(Succeed())
+	g.Expect(mt.Name).To(Equal(originalName))
+}
+
+func TestControlPlaneUpdateImmutableObjectNamesNoKubeadmControlPlane(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+	client := test.NewFakeKubeClient()
+	cp := controlPlaneStackedEtcd()
+	originalCPMachineTemplateName := "my-machine-template-1"
+	cp.ControlPlaneMachineTemplate.Name = originalCPMachineTemplateName
+
+	g.Expect(cp.UpdateImmutableObjectNames(ctx, client, dummyRetriever, noChangesCompare)).To(Succeed())
+	g.Expect(cp.ControlPlaneMachineTemplate.Name).To(Equal(originalCPMachineTemplateName))
+}
+
+func TestControlPlaneUpdateImmutableObjectNamesErrorReadingControlPlane(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+	cp := controlPlaneStackedEtcd()
+	client := test.NewFakeKubeClientAlwaysError()
+
+	g.Expect(
+		cp.UpdateImmutableObjectNames(ctx, client, dummyRetriever, noChangesCompare),
+	).To(
+		MatchError(ContainSubstring("reading current kubeadm control plane from API")),
+	)
+}
+
+func TestControlPlaneUpdateImmutableObjectNamesErrorUpdatingName(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+	cp := controlPlaneStackedEtcd()
+	originalCPMachineTemplateName := "my-machine-template"
+	cp.ControlPlaneMachineTemplate.Name = originalCPMachineTemplateName
+	cp.KubeadmControlPlane.Spec.MachineTemplate.InfrastructureRef.Name = originalCPMachineTemplateName
+	client := test.NewFakeKubeClient(clientutil.ObjectsToClientObjects(cp.Objects())...)
+
+	g.Expect(
+		cp.UpdateImmutableObjectNames(ctx, client, dummyRetriever, withChangesCompare),
+	).To(
+		MatchError(ContainSubstring("incrementing name for DockerMachineTemplate eksa-system/my-machine-template")),
+	)
+}
+
+func TestControlPlaneUpdateImmutableObjectNamesSuccessStackedEtcd(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+	cp := controlPlaneStackedEtcd()
+	originalCPMachineTemplateName := "my-machine-template-1"
+	cp.ControlPlaneMachineTemplate.Name = originalCPMachineTemplateName
+	cp.KubeadmControlPlane.Spec.MachineTemplate.InfrastructureRef.Name = originalCPMachineTemplateName
+	client := test.NewFakeKubeClient(clientutil.ObjectsToClientObjects(cp.Objects())...)
+
+	g.Expect(cp.UpdateImmutableObjectNames(ctx, client, dummyRetriever, noChangesCompare)).To(Succeed())
+}
+
+func TestControlPlaneUpdateImmutableObjectNamesNoEtcdCluster(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+	cp := controlPlaneStackedEtcd()
+	originalCPMachineTemplateName := "my-machine-template-1"
+	cp.ControlPlaneMachineTemplate.Name = originalCPMachineTemplateName
+	cp.KubeadmControlPlane.Spec.MachineTemplate.InfrastructureRef.Name = originalCPMachineTemplateName
+	client := test.NewFakeKubeClient(clientutil.ObjectsToClientObjects(cp.Objects())...)
+	cp.EtcdCluster = etcdCluster()
```

Comment:
Can you explain what this test is representing? Are we mimicking a case where the etcdCluster was not originally a part of the eksa cluster so it ignores it since we don't allow configuring etcd during upgrade?

Reply:
This is dealing with the scenario where an etcdCluster is defined in the CP spec but that etcdCluster doesn't exist yet "in" the api server.

That scenario might it might no be possible depending on preflight validations, but this code makes no assumption about those. The logic followed is: if there is not an EtcdCluster object yet, that means there won't be a name collision, hence we don't need to generate a new name for the machine template.